File: engine/factory.py
"""
Factory definitions for Mosaic-AI.

Mosaic-AI hat:
  - 4 kleine Fabriken (je 4 Steine auf der Sun-Seite)
  - 1 große Fabrik   (5 Steine auf der Sun-Seite + Moon-Pool für Reste)

Sun-Mechanik (kleine & große Fabrik):
  Spieler nimmt ALLE Steine einer Farbe von der Sun-Seite.
  Die restlichen Steine legt er in gewählter Reihenfolge auf die Moon-Seite.

Moon-Mechanik (nur kleine Fabriken):
  Spieler nimmt alle TOP-Steine einer Farbe von den Moon-Stapeln.

Moon-Pool (große Fabrik):
  Flacher Pool für Steine die direkt in der großen Fabrik übrig bleiben
  (nach Sun-Entnahme). Steine kleiner Fabriken bleiben auf deren eigenen
  Moon-Stapeln — sie wandern NICHT in den Moon-Pool der großen Fabrik.
  Spieler nimmt alle Steine einer Farbe aus dem Pool.

Der Startspieler-Marker liegt bei der großen Fabrik.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional

from engine.tile import TileColor
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from engine.dome import BonusChip


@dataclass
class Factory:
    """Eine kleine Fabrik."""
    factory_id:           int
    sun_tiles:            list[TileColor]        = field(default_factory=list)
    moon_stacks:          list[list[TileColor]]  = field(default_factory=list)
    bonus_chip:           Optional[object]       = None   # BonusChip | None (verdeckt)
    bonus_chip_revealed:  bool                   = False  # True sobald Fabrik leer

    # ------------------------------------------------------------------
    # Sun-Seite
    # ------------------------------------------------------------------

    @property
    def sun_is_empty(self) -> bool:
        return len(self.sun_tiles) == 0

    def sun_colors(self) -> set[TileColor]:
        return set(self.sun_tiles)

    def take_from_sun(self, color: TileColor) -> tuple[list[TileColor], list[TileColor]]:
        """
        Nimmt alle Steine der gewählten Farbe von der Sun-Seite.
        Gibt (genommene, übrige) zurück — übrige müssen auf Moon gelegt werden.
        """
        if color not in self.sun_colors():
            raise ValueError(
                f"Farbe {color.value} nicht auf Sun-Seite von Fabrik {self.factory_id}."
            )
        taken     = [t for t in self.sun_tiles if t == color]
        remaining = [t for t in self.sun_tiles if t != color]
        self.sun_tiles = []
        return taken, remaining

    def place_on_moon(self, ordered_tiles: list[TileColor]) -> None:
        """
        Legt alle verbleibenden Fliesen als EINEN Stapel auf die Moon-Seite.
        ordered_tiles: Reihenfolge vom Spieler gewählt — Index 0 = unten, letzter = oben (sichtbar).
        """
        if ordered_tiles:
            self.moon_stacks.append(list(ordered_tiles))

    # ------------------------------------------------------------------
    # Moon-Seite
    # ------------------------------------------------------------------

    @property
    def moon_is_empty(self) -> bool:
        return len(self.moon_stacks) == 0

    def moon_top_colors(self) -> set[TileColor]:
        return {stack[-1] for stack in self.moon_stacks if stack}

    def take_from_moon(self, color: TileColor) -> list[TileColor]:
        """
        Nimmt alle TOP-Steine der gewählten Farbe von den Moon-Stapeln.
        """
        if color not in self.moon_top_colors():
            raise ValueError(
                f"Farbe {color.value} nicht oben auf Moon-Stapeln von Fabrik {self.factory_id}."
            )
        taken = []
        surviving = []
        for stack in self.moon_stacks:
            if stack and stack[-1] == color:
                taken.append(stack.pop())
                if stack:
                    surviving.append(stack)
            else:
                surviving.append(stack)
        self.moon_stacks = surviving

        if self.sun_is_empty and self.moon_is_empty:
            self.bonus_chip_revealed = True

        return taken

    @property
    def is_fully_empty(self) -> bool:
        return self.sun_is_empty and self.moon_is_empty

    def __repr__(self) -> str:
        sun  = [c.value for c in self.sun_tiles]
        moon = [[c.value for c in s] for s in self.moon_stacks]
        chip = repr(self.bonus_chip) if self.bonus_chip_revealed and self.bonus_chip else ("[verdeckt]" if self.bonus_chip else "kein")
        return (
            f"Factory(id={self.factory_id}, sun={sun}, moon={moon}, "
            f"bonus={chip})"
        )


@dataclass
class LargeFactory:
    """
    Die große Fabrik (zentrales Display).

    Sun-Seite:  5 Steine zu Rundenbeginn, offen ausgelegt.
    Moon-Pool:  flacher Pool, Reste von kleinen Fabriken landen hier.

    Der Startspieler-Marker liegt hier. Wer Steine nimmt während
    has_first_player_marker=True ist, erhält den Marker.
    """
    sun_tiles:               list[TileColor] = field(default_factory=list)
    moon_pool:               list[TileColor] = field(default_factory=list)
    has_first_player_marker: bool            = True

    # ------------------------------------------------------------------
    # Sun-Seite
    # ------------------------------------------------------------------

    @property
    def sun_is_empty(self) -> bool:
        return len(self.sun_tiles) == 0

    def sun_colors(self) -> set[TileColor]:
        return set(self.sun_tiles)

    def take_from_sun(self, color: TileColor) -> tuple[list[TileColor], list[TileColor], bool]:
        """
        Nimmt alle Steine einer Farbe von der Sun-Seite.
        Gibt (genommene, übrige, hatte_marker) zurück.
        Übrige wandern in den Moon-Pool (Aufgabe des Aufrufers).
        """
        if color not in self.sun_colors():
            raise ValueError(f"Farbe {color.value} nicht auf Sun-Seite der großen Fabrik.")
        taken     = [t for t in self.sun_tiles if t == color]
        remaining = [t for t in self.sun_tiles if t != color]
        self.sun_tiles = []
        marker = self.has_first_player_marker
        self.has_first_player_marker = False
        return taken, remaining, marker

    # ------------------------------------------------------------------
    # Moon-Pool
    # ------------------------------------------------------------------

    @property
    def moon_is_empty(self) -> bool:
        return len(self.moon_pool) == 0

    def moon_colors(self) -> set[TileColor]:
        return set(self.moon_pool)

    def take_from_moon(self, color: TileColor) -> tuple[list[TileColor], bool]:
        """
        Nimmt alle Steine einer Farbe aus dem Moon-Pool.
        Gibt (genommene, hatte_marker) zurück.
        """
        if color not in self.moon_colors():
            raise ValueError(f"Farbe {color.value} nicht im Moon-Pool der großen Fabrik.")
        taken = [t for t in self.moon_pool if t == color]
        self.moon_pool = [t for t in self.moon_pool if t != color]
        marker = self.has_first_player_marker
        self.has_first_player_marker = False
        return taken, marker

    def add_to_moon(self, tiles: list[TileColor]) -> None:
        """Fügt Steine zum Moon-Pool hinzu (Reste von kleinen Fabriken)."""
        self.moon_pool.extend(tiles)

    # ------------------------------------------------------------------
    # Allgemein
    # ------------------------------------------------------------------

    @property
    def is_empty(self) -> bool:
        return self.sun_is_empty and self.moon_is_empty and not self.has_first_player_marker

    def reset_for_new_round(self) -> None:
        self.sun_tiles = []
        self.moon_pool = []
        self.has_first_player_marker = True

    def __repr__(self) -> str:
        sun  = [c.value for c in self.sun_tiles]
        moon = [c.value for c in self.moon_pool]
        marker = " ★" if self.has_first_player_marker else ""
        return f"LargeFactory(sun={sun}, moon={moon}{marker})"

File: engine/src/factory.rs
//! Kleine Fabriken + große Fabrik (Tischmitte) — Port von engine/factory.py.

use crate::dome::BonusChip;
use crate::tile::TileColor;

/// Eine kleine Fabrik.
#[derive(Debug, Clone)]
pub struct Factory {
    pub factory_id: usize,
    pub sun_tiles: Vec<TileColor>,
    pub moon_stacks: Vec<Vec<TileColor>>,
    pub bonus_chip: Option<BonusChip>,
    pub bonus_chip_revealed: bool,
}

impl Factory {
    pub fn new(factory_id: usize) -> Self {
        Factory {
            factory_id,
            sun_tiles: Vec::new(),
            moon_stacks: Vec::new(),
            bonus_chip: None,
            bonus_chip_revealed: false,
        }
    }

    pub fn sun_is_empty(&self) -> bool {
        self.sun_tiles.is_empty()
    }

    pub fn sun_colors(&self) -> Vec<TileColor> {
        let mut v: Vec<TileColor> = Vec::new();
        for &t in &self.sun_tiles {
            if !v.contains(&t) {
                v.push(t);
            }
        }
        v
    }

    /// Nimmt alle Steine der Farbe von der Sun-Seite. Gibt (genommene, übrige) zurück.
    pub fn take_from_sun(
        &mut self,
        color: TileColor,
    ) -> Result<(Vec<TileColor>, Vec<TileColor>), String> {
        if !self.sun_tiles.contains(&color) {
            return Err(format!(
                "Farbe {} nicht auf Sun-Seite von Fabrik {}.",
                color.value(),
                self.factory_id
            ));
        }
        let taken: Vec<TileColor> = self.sun_tiles.iter().copied().filter(|&t| t == color).collect();
        let remaining: Vec<TileColor> =
            self.sun_tiles.iter().copied().filter(|&t| t != color).collect();
        self.sun_tiles.clear();
        Ok((taken, remaining))
    }

    /// Legt die übrigen Fliesen als EINEN Stapel auf die Moon-Seite
    /// (Index 0 = unten, letzter = oben/sichtbar).
    pub fn place_on_moon(&mut self, ordered_tiles: Vec<TileColor>) {
        if !ordered_tiles.is_empty() {
            self.moon_stacks.push(ordered_tiles);
        }
    }

    pub fn moon_is_empty(&self) -> bool {
        self.moon_stacks.is_empty()
    }

    /// Top-Farben der Moon-Stapel.
    pub fn moon_top_colors(&self) -> Vec<TileColor> {
        let mut v: Vec<TileColor> = Vec::new();
        for stack in &self.moon_stacks {
            if let Some(&top) = stack.last() {
                if !v.contains(&top) {
                    v.push(top);
                }
            }
        }
        v
    }

    /// Nimmt alle TOP-Steine der Farbe von den Moon-Stapeln.
    pub fn take_from_moon(&mut self, color: TileColor) -> Result<Vec<TileColor>, String> {
        if !self.moon_top_colors().contains(&color) {
            return Err(format!(
                "Farbe {} nicht oben auf Moon-Stapeln von Fabrik {}.",
                color.value(),
                self.factory_id
            ));
        }
        let mut taken = Vec::new();
        let mut surviving: Vec<Vec<TileColor>> = Vec::new();
        for mut stack in std::mem::take(&mut self.moon_stacks) {
            if stack.last() == Some(&color) {
                taken.push(stack.pop().unwrap());
                if !stack.is_empty() {
                    surviving.push(stack);
                }
            } else {
                surviving.push(stack);
            }
        }
        self.moon_stacks = surviving;
        Ok(taken)
    }

    pub fn is_fully_empty(&self) -> bool {
        self.sun_is_empty() && self.moon_is_empty()
    }
}

/// Die große Fabrik (Tischmitte). Trägt den Startspieler-Marker.
#[derive(Debug, Clone)]
pub struct LargeFactory {
    pub sun_tiles: Vec<TileColor>,
    pub moon_pool: Vec<TileColor>,
    pub has_first_player_marker: bool,
    /// Regelbuch S.10: konnten Beutel+Turm keine 2 verschiedenen Farben mehr
    /// liefern, wird die monochrome Befüllung akzeptiert -- dann (und NUR
    /// dann) vergibt bereits die Sonnen-Nahme die Startspielerfliese, weil
    /// der Mondbereich sonst leer bliebe und der Marker unnehmbar wäre.
    pub monochrome_fallback: bool,
}

impl Default for LargeFactory {
    fn default() -> Self {
        LargeFactory {
            sun_tiles: Vec::new(),
            moon_pool: Vec::new(),
            has_first_player_marker: true,
            monochrome_fallback: false,
        }
    }
}

impl LargeFactory {
    pub fn sun_is_empty(&self) -> bool {
        self.sun_tiles.is_empty()
    }

    pub fn sun_colors(&self) -> Vec<TileColor> {
        let mut v: Vec<TileColor> = Vec::new();
        for &t in &self.sun_tiles {
            if !v.contains(&t) {
                v.push(t);
            }
        }
        v
    }

    /// Nimmt alle Steine einer Farbe von der Sun-Seite.
    /// Gibt (genommene, übrige, hatte_marker) zurück; übrige → Moon-Pool (Aufrufer).
    pub fn take_from_sun(
        &mut self,
        color: TileColor,
    ) -> Result<(Vec<TileColor>, Vec<TileColor>, bool), String> {
        if !self.sun_tiles.contains(&color) {
            return Err(format!(
                "Farbe {} nicht auf Sun-Seite der großen Fabrik.",
                color.value()
            ));
        }
        let taken: Vec<TileColor> = self.sun_tiles.iter().copied().filter(|&t| t == color).collect();
        let remaining: Vec<TileColor> =
            self.sun_tiles.iter().copied().filter(|&t| t != color).collect();
        self.sun_tiles.clear();
        // Regelbuch S.5: NUR die erste Nahme vom MONDbereich vergibt die
        // Startspielerfliese -- die Sonnen-Nahme lässt den Marker liegen.
        // Einzige Ausnahme (Regelbuch S.10): monochrome Notbefüllung, dann
        // bleibt der Mond leer und der Marker geht mit den 5 gleichfarbigen.
        let marker = self.monochrome_fallback && self.has_first_player_marker;
        if marker {
            self.has_first_player_marker = false;
        }
        Ok((taken, remaining, marker))
    }

    pub fn moon_is_empty(&self) -> bool {
        self.moon_pool.is_empty()
    }

    pub fn moon_colors(&self) -> Vec<TileColor> {
        let mut v: Vec<TileColor> = Vec::new();
        for &t in &self.moon_pool {
            if !v.contains(&t) {
                v.push(t);
            }
        }
        v
    }

    /// Nimmt alle Steine einer Farbe aus dem Moon-Pool. Gibt (genommene, hatte_marker).
    pub fn take_from_moon(&mut self, color: TileColor) -> Result<(Vec<TileColor>, bool), String> {
        if !self.moon_pool.contains(&color) {
            return Err(format!(
                "Farbe {} nicht im Moon-Pool der großen Fabrik.",
                color.value()
            ));
        }
        let taken: Vec<TileColor> = self.moon_pool.iter().copied().filter(|&t| t == color).collect();
        self.moon_pool.retain(|&t| t != color);
        let marker = self.has_first_player_marker;
        self.has_first_player_marker = false;
        Ok((taken, marker))
    }

    pub fn add_to_moon(&mut self, tiles: &[TileColor]) {
        self.moon_pool.extend_from_slice(tiles);
    }

    pub fn is_empty(&self) -> bool {
        self.sun_is_empty() && self.moon_is_empty() && !self.has_first_player_marker
    }

    pub fn reset_for_new_round(&mut self) {
        self.sun_tiles.clear();
        self.moon_pool.clear();
        self.has_first_player_marker = true;
        self.monochrome_fallback = false;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tile::TileColor::*;

    #[test]
    fn sun_take_splits_and_remainder() {
        let mut f = Factory::new(1);
        f.sun_tiles = vec![Rot, Rot, Blau, Gelb];
        let (taken, rest) = f.take_from_sun(Rot).unwrap();
        assert_eq!(taken, vec![Rot, Rot]);
        assert_eq!(rest, vec![Blau, Gelb]);
        assert!(f.sun_is_empty());
        f.place_on_moon(rest);
        assert_eq!(f.moon_top_colors(), vec![Gelb]); // letzter = oben
    }

    #[test]
    fn moon_take_top_only() {
        let mut f = Factory::new(1);
        f.moon_stacks = vec![vec![Blau, Rot], vec![Gelb, Rot], vec![Schwarz]];
        let taken = f.take_from_moon(Rot).unwrap();
        assert_eq!(taken.len(), 2); // beide Top-Rot
        // Reste: [Blau], [Gelb], [Schwarz]
        assert_eq!(f.moon_stacks.len(), 3);
        assert!(f.take_from_moon(Rot).is_err()); // kein Rot mehr oben
    }

    #[test]
    fn large_factory_marker_taken_once() {
        // R2 (Vollaudit 2026-07-21): der Marker geht mit der ERSTEN
        // Mond-Nahme, nicht mit der Sonnen-Nahme.
        let mut lf = LargeFactory::default();
        lf.moon_pool = vec![Rot, Blau, Blau];
        let (taken, marker) = lf.take_from_moon(Blau).unwrap();
        assert_eq!(taken.len(), 2);
        assert!(marker);
        assert!(!lf.has_first_player_marker);
        // Zweite Mond-Nahme: Marker ist weg.
        let (taken2, marker2) = lf.take_from_moon(Rot).unwrap();
        assert_eq!(taken2.len(), 1);
        assert!(!marker2);
    }

    #[test]
    fn large_factory_sun_take_leaves_marker() {
        // R2: die Sonnen-Nahme lässt den Marker liegen -- erst die
        // anschließende erste Mond-Nahme holt ihn.
        let mut lf = LargeFactory::default();
        lf.sun_tiles = vec![Rot, Blau, Blau];
        let (taken, rest, marker) = lf.take_from_sun(Blau).unwrap();
        assert_eq!(taken.len(), 2);
        assert_eq!(rest, vec![Rot]);
        assert!(!marker, "Sonnen-Nahme darf den Marker nicht vergeben");
        assert!(lf.has_first_player_marker);
        lf.add_to_moon(&rest);
        let (_, moon_marker) = lf.take_from_moon(Rot).unwrap();
        assert!(moon_marker, "erste Mond-Nahme holt den Marker");
        assert!(!lf.has_first_player_marker);
        assert!(lf.is_empty());
    }

    #[test]
    fn large_factory_monochrome_fallback_gives_marker_on_sun_take() {
        // R3: bei monochromer Notbefüllung (Regelbuch S.10) vergibt
        // ausnahmsweise die Sonnen-Nahme den Marker.
        let mut lf = LargeFactory::default();
        lf.sun_tiles = vec![Rot, Rot, Rot, Rot, Rot];
        lf.monochrome_fallback = true;
        let (taken, rest, marker) = lf.take_from_sun(Rot).unwrap();
        assert_eq!(taken.len(), 5);
        assert!(rest.is_empty());
        assert!(marker, "monochromer Fallback: Sonnen-Nahme vergibt den Marker");
        assert!(!lf.has_first_player_marker);
        assert!(lf.is_empty());
    }
}
